File: design-algo/HW2_merge_sort.py
# =============================================================
# HOMEWORK 2 — Example of Merge Sort (Python version)
#
# Task: Sort the array
#       [18, 26, 32, 6, 43, 15, 9, 1, 22, 26, 19, 55, 37, 43, 99, 2]
#       using the Merge Sort algorithm and count the number of steps.
#
# For n = 16 elements: log2(16) = 4 merge levels.
#
# HOW TO RUN:
#   python3 HW2_merge_sort.py
# =============================================================

step_count = 0         


def merge(arr: list[int], left: int, mid: int, right: int) -> None:
    global step_count

    left_half  = arr[left : mid + 1]  
    right_half = arr[mid + 1 : right + 1] 

    i = j = 0
    k = left

    while i < len(left_half) and j < len(right_half):
        if left_half[i] <= right_half[j]:
            arr[k] = left_half[i]
            i += 1
        else:
            arr[k] = right_half[j]
            j += 1
        k += 1
        step_count += 1

    while i < len(left_half):
        arr[k] = left_half[i]
        i += 1
        k += 1
        step_count += 1

    while j < len(right_half):
        arr[k] = right_half[j]
        j += 1
        k += 1
        step_count += 1


def merge_sort(arr: list[int], left: int, right: int) -> None:
    if left < right:
        mid = left + (right - left) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)


def main() -> None:
    arr = [18, 26, 32, 6, 43, 15, 9, 1, 22, 26, 19, 55, 37, 43, 99, 2]
    n   = len(arr)

    print("=== HOMEWORK 2: Merge Sort ===\n")
    print(f"Original array ({n} elements):")
    print(arr)

    merge_sort(arr, 0, n - 1)

    print("\nSorted array:")
    print(arr)

    import math
    levels = int(math.log2(n))
    print(f"\nNumber of merge levels (log2({n})): {levels}")
    print(f"Total comparison/copy steps counted: {step_count}")


if __name__ == "__main__":
    main()

File: design-algo/HW2_merge_sort.java
// =============================================================
// HOMEWORK 2 — Example of Merge Sort (Java version)
//
// Task: Sort the array
//       [18, 26, 32, 6, 43, 15, 9, 1, 22, 26, 19, 55, 37, 43, 99, 2]
//       using the Merge Sort algorithm and count the number of steps.
//
// For n = 16 elements: log2(16) = 4 merge levels.
//
// HOW TO COMPILE AND RUN:
//   javac HW2_merge_sort.java
//   java  HW2_merge_sort
// =============================================================

import java.util.Arrays;

public class HW2_merge_sort {

    private static int stepCount = 0;

    // ----- merge two sorted halves back into arr[left..right] -----
    private static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = Arrays.copyOfRange(arr, left, left + n1);
        int[] R = Arrays.copyOfRange(arr, mid + 1, mid + 1 + n2);

        int i = 0, j = 0, k = left;

        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) arr[k++] = L[i++];
            else               arr[k++] = R[j++];
            stepCount++;
        }
        while (i < n1) { arr[k++] = L[i++]; stepCount++; }
        while (j < n2) { arr[k++] = R[j++]; stepCount++; }
    }

    // ----- recursive divide-and-conquer -----
    private static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }

    // ----- main -----
    public static void main(String[] args) {
        int[] arr = {18, 26, 32, 6, 43, 15, 9, 1, 22, 26, 19, 55, 37, 43, 99, 2};
        int n = arr.length;

        System.out.println("=== HOMEWORK 2: Merge Sort ===\n");
        System.out.println("Original array (" + n + " elements):");
        System.out.println(Arrays.toString(arr));

        mergeSort(arr, 0, n - 1);

        System.out.println("\nSorted array:");
        System.out.println(Arrays.toString(arr));

        int levels = (int)(Math.log(n) / Math.log(2));
        System.out.println("\nNumber of merge levels (log2(" + n + ")): " + levels);
        System.out.println("Total comparison/copy steps counted: " + stepCount);
    }
}
